ethcore/src/client/client.rs

```diff
@@ -371,9 +371,24 @@ impl<V> Client<V> where V: Verifier {
 			return Some(self.state())
 		}
 
-		self.block_header(id).map(|header| {
+		let block_number = match self.block_number(id.clone()) {
+			Some(num) => num,
+			None => return None,
+		};
+
+		self.block_header(id).and_then(|header| {
 			let db = self.state_db.lock().unwrap().boxed_clone();
-			State::from_existing(db, HeaderView::new(&header).state_root(), self.engine.account_start_nonce())
+
+			let root = HeaderView::new(&header).state_root();
+			// check that the block is not too old -- blocks within `HISTORY` blocks of the best will
+			// always be available. If the block could be too old, check if its state root is valid.
+			if db.does_pruning()
+			&& self.chain.best_block_number() >= block_number + HISTORY
```

Comment:
bad style - run-on lines should be indented. in this case, i would place `let is_old = self.chain.best_block_number() >= block_number + HISTORY;` on its own line above and keep the `if` condition on a single line.